Add guide for YOLOv7 ONNX inference

Y-T-G opened the pull request:
Adds a guide to transform the ONNX model from YOLOv7 to Ultralytics compatible format, making it possible to load and run inference with it in Ultralytics. TensorRT inference is also supported.


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
YOLOv7 models can now be used for ONNX and TensorRT inference in Ultralytics, with detailed documentation and scripts to help users export and modify models for compatibility. 🚀

### 📊 Key Changes
- Updated documentation to clarify that only ONNX and TensorRT inference is supported for YOLOv7 in Ultralytics.
- Added step-by-step guides for exporting YOLOv7 models to ONNX and TensorRT formats.
- Provided a Python script to modify YOLOv7 ONNX models for Ultralytics compatibility.
- Included example code for running inference with YOLOv7 ONNX and TensorRT models in Ultralytics.
- Clarified that training YOLOv7 models must still be done using the original YOLOv7 repository.

### 🎯 Purpose & Impact
- Enables users to run YOLOv7 inference directly in Ultralytics using ONNX or TensorRT models, expanding supported model options. 🖼️
- Makes it easier for users to leverage YOLOv7’s speed and accuracy within the Ultralytics ecosystem.
- Provides clear, practical instructions for exporting and preparing YOLOv7 models, reducing confusion and setup time. ⏱️
- Ensures users understand the current limitations (inference only, no training) and how to work around them.

Y-T-G committed:
Remove YOLOv7 from supported models

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

UltralyticsAssistant commented:
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description and ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running the appropriate commands locally.
- ✅ **Ensure CI Checks Pass**: Verify all Continuous Integration (CI) checks are passing. If any checks fail, please address the issues to proceed.
- ✅ **Update Documentation**: Update the relevant documentation for any new or modified features to keep our resources clear and consistent.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our Contributor License Agreement (CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message below the PR.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For additional guidance, please refer to our contributing documentation or leave a comment if you have any questions. This is an automated response, but an Ultralytics engineer will review and assist with your PR soon. Thank you for contributing to Ultralytics! 🚀

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Y-T-G commented:
@Laughing-q Would it be possible to add inference support for YOLOv7 using exported format like ONNX and engine? The exported format could work as long as we get the output to be similar to the other Ultralytics YOLO models.

glenn-jocher committed:
Merge branch 'main' into models-docs

glenn-jocher commented:
@Y-T-G I think better to add a note on the YOLOv7 model page like we have here for Explorer https://docs.ultralytics.com/datasets/explorer/

Y-T-G committed:
Support YOLOv7 ONNX inference

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Y-T-G commented:
@glenn-jocher I updated the guide and added instruction to run YOLOv7 ONNX inference with Ultralytics instead. So now we can say we have partial support.

UltralyticsAssistant commented:
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description and ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Continuous Integration (CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant documentation for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our Contributor License Agreement (CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For additional guidance, refer to the project's contributing guidelines or leave a comment if you have any questions. An Ultralytics engineer will review your PR soon and provide feedback if needed. Thank you for contributing to Ultralytics! 🚀

glenn-jocher committed:
Merge branch 'main' into models-docs

Laughing-q commented:
@Y-T-G Hey nice idea here! I think by default the `--end2end` option in yolov7 repo is introducing the nms module into the onnx model,
```bash
python export.py --weights yolov7-tiny.pt --grid --end2end --simplify --topk-all 100 --iou-thres 0.65 --conf-thres 0.35 --img-size 640 640 --max-wh 640
```
which I think the subsequent manipulation on onnx would be much easier if we eliminate this option?
The output I got without the option:
![pic-241231-1651-12](https://github.com/user-attachments/assets/45c6a521-a1d3-483a-9150-8efc4d30aa9f)
based on this I think we just need to add a simple `Transpose` op at the end to update `1,25200,85]` to `1,85,25200]`?

Y-T-G committed:
Add TensorRT export instructions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Y-T-G commented:
@Laughing-q We could do that. It should make the ONNX export script much smaller. And probably even compatible with other formats. I just added instruction for TensorRT as well.

Y-T-G commented:
@Laughing-q It works. But I also have to multiply the objectness score with class scores in the ONNX graph and drop the objectness score after that to make it aligned with Ultralytics which makes the script longer. And it's also somewhat slower when compared to the NMS embedded version.

**YOLOv7-tiny TensorRT with NMS**
```python
%timeit results = model(img, verbose=False)
5.66 ms ± 666 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**YOLOv7-tiny TensorRT without NMS**
```python
%timeit results = model(img, verbose=False)
6.72 ms ± 800 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

glenn-jocher committed:
Merge branch 'main' into models-docs

Laughing-q commented:
@Y-T-G ah I see I forgot about the legacy `objectness`, then probably it's better we stick with the nms implementation.

Y-T-G committed:
Merge branch 'main' into models-docs

Y-T-G commented:
@glenn-jocher This can be merged.

UltralyticsAssistant committed:
Merge branch 'main' into models-docs

Laughing-q commented on ultralytics/nn/autobackend.py:
```diff
@@ -225,17 +225,21 @@ def __init__(
-                        buffer_ptr=y_tensor.data_ptr(),
-                    )
-                    bindings.append(y_tensor)
+                try:
```
@Y-T-G Do we still need this change? I tested the exported and modified yolov7-ultralytics.onnx model without this change and it still works to me.

Y-T-G replied:
It's not required after I added zero padding to make the output shape fixed.

Laughing-q replied:
@Y-T-G nice then let's remove this!

Y-T-G committed:
Update autobackend.py

Y-T-G committed:
Update autobackend.py

UltralyticsAssistant committed:
Merge branch 'main' into models-docs

Laughing-q committed:
Merge branch 'main' into models-docs

Laughing-q commented:
@glenn-jocher this one is good to go

Laughing-q committed:
Merge branch 'main' into models-docs

UltralyticsAssistant committed:
Merge branch 'main' into models-docs

UltralyticsAssistant committed:
Merge branch 'main' into models-docs

Y-T-G committed:
Merge branch 'main' into models-docs

Signed-off-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>

Y-T-G committed:
Merge 5667a4244009c922ca6730c2df8d9d154a34e678 into 20196b57d0285856d2b9e9ef095beb39d0643c8e

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

Y-T-G committed:
Merge 3fc1a24522b8c085a7ae6704ca636ad61ae2db3b into 20196b57d0285856d2b9e9ef095beb39d0643c8e

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into models-docs

UltralyticsAssistant committed:
Merge branch 'main' into models-docs

UltralyticsAssistant committed:
Merge branch 'main' into models-docs

RizwanMunawar committed:
Merge branch 'main' into models-docs

UltralyticsAssistant commented:
🎉 Fantastic work, @Y-T-G, with valuable contributions from @glenn-jocher, @Laughing-q, and @RizwanMunawar! Merging this PR is a big step forward for the Ultralytics community—expanding YOLOv7 ONNX and TensorRT inference support and making it more accessible than ever.

As Henry Ford once said, "Coming together is a beginning, staying together is progress, and working together is success." This collaboration truly exemplifies that spirit—empowering more users to harness YOLOv7’s capabilities within Ultralytics with clarity and ease.

Thank you all for your dedication and thoughtful documentation. Your efforts strengthen the Ultralytics ecosystem and inspire the entire community! 🚀
